src/test/ui/rfc-2091-track-caller/tracked-trait-obj.rs

```diff
@@ -2,22 +2,60 @@
 
 trait Tracked {
     #[track_caller]
-    fn handle(&self) {
+    fn track_caller_trait_method(&self, line: u32, col: u32) {
         let location = std::panic::Location::caller();
         assert_eq!(location.file(), file!());
-        // we only call this via trait object, so the def site should *always* be returned
-        assert_eq!(location.line(), line!() - 4);
-        assert_eq!(location.column(), 5);
+        // The trait method definition is annotated with `#[track_caller]`,
+        // so caller location information will work through a method
+        // call on a trait object
+        assert_eq!(location.line(), line, "Bad line");
+        assert_eq!(location.column(), col, "Bad col");
     }
+
+    fn track_caller_not_on_trait_method(&self);
+
+    #[track_caller]
+    fn track_caller_through_self(self: Box<Self>, line: u32, col: u32);
 }
 
-impl Tracked for () {}
-impl Tracked for u8 {}
+impl Tracked for () {
+    // We have `#[track_caller]` on the implementation of the method,
+    // but not on the definition of the method in the trait. Therefore,
+    // caller location information will *not* work through a method call
+    // on a trait object. Instead, we will get the location of this method
+    #[track_caller]
+    fn track_caller_not_on_trait_method(&self) {
+        let location = std::panic::Location::caller();
+        assert_eq!(location.file(), file!());
+        assert_eq!(location.line(), line!() - 3);
+        assert_eq!(location.column(), 5);
+    }
```

Comment:
Is there any reason why the `file!`/`line!` don't fall-back to their usual behaviour of providing precisely the line/column they are on, rather than that of where the method definition item begins? As implemented right now, this seems like a 3rd new behaviour for these macros.

(Though with the behaviour as it is, I'm sure that the bikeshed could go on ad infinitum here)

Reply:
This is the current behavior whenever we need to generate a shim:

```rust
trait Foo {
    #[track_caller] fn bar(&self);
}

impl Foo for () {
    fn bar(&self) {
        panic!("Where does this end up?")
    }
}

fn foo(val: &dyn Foo) {
    val.bar();
}

fn main() {
    foo(&());
}
```

On the latest nightly, this prints `thread 'main' panicked at 'Where does this end up?', src/main.rs:6:5`

I agree that it would probably be good to change this, but I think that could be done separately.